### PR TITLE
Add support for configuration precedence

### LINKED ISSUE #3545: Use result type for `Lookup`
**Describe the current behavior**

When looking up configuration, we currently return a specific `IsNotConfiguredError` when there's no configuration to return.

As [noted](https://github.com/Azure/azure-service-operator/pull/3469#discussion_r1388846464) by #matthchr this is capturing a bunch of stack information that's not required, which will be contributing to GC churn, and slowing performance, amongst other issues.

**Describe the improvement**

We could introduce a `LookupResult` struct that captures the result of the lookup - containing either a value or an indication that nothing was configured, leaving the `error` return for actual problems.

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Modifies our configuration model to allow for precedence - where you can configure an item to apply to an entire group or type, but override it at a lower level. Driven by the need to specify `PayloadType` at multiple levels of granularity.

Also moves the configuration API to a more idiomatic style where the various `Lookup()` methods return `(T, bool)`, making consumption simpler.

Closes #3545, though with a different approach

**Special notes for your reviewer**:

* Removes use of `NotConfiguredError`
* Changes `configurable[T].Lookup()` to return (T, bool)
* Modifies `groupAccess` to allow use of a type override, `groupAccess.Lookup()` to return `(T, bool)`, and relocates to a separate file
* Modifies `typeAccess` to allow use of a property override, `typeAccess.Lookup()` to return `(T, bool)`, and relocates to a separate file
* Modifies `propertyAccess.Lookup()` to return `(T, bool)`, and relocates to a separate file

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

**Comment by theunrepentantgeek** [comment]
After this has merged, configuration for `PayloadType` will need to change as follows:

On `ObjectModelConfiguration`, change the field type:

``` go
	// Multi-level access fields here (alphabetical, please)
	PayloadType propertyAccess[PayloadType]
```

with initialization to look like this:

``` go
	// Initialize multi-level access fields here (alphabetical, please)
	result.PayloadType = makeGroupAccess[PayloadType](
		result,
		func(c *GroupConfiguration) *configurable[PayloadType] { return &c.PayloadType },
	).withTypeOverride(
		func(c *TypeConfiguration) *configurable[PayloadType] { return &c.PayloadType },
	).withPropertyOverride(
		func(c *PropertyConfiguration) *configurable[PayloadType] { return &c.PayloadType },
	)
```

Consumption will need to be updated as well.

File: v2/tools/generator/internal/config/property_configuration.go [hunk 1]
```diff
@@ -27,8 +27,9 @@ type PropertyConfiguration struct {
 	ImportConfigMapMode            configurable[ImportConfigMapMode] // The config map mode
 	IsSecret                       configurable[bool]                // Specify whether this property is a secret
 	NameInNextVersion              configurable[string]              // Name this property has in the next version
-	RenameTo                       configurable[string]              // Name this property should be renamed to
-	ResourceLifecycleOwnedByParent configurable[string]              // Name of the parent resource which owns the lifecycle of the sub-resource.
+	PayloadType                    configurable[PayloadType]
```
Comment (matthchr) [inline v2/tools/generator/internal/config/property_configuration.go:30]: Deserves a comment as well (similar to others?)

**Review by matthchr — APPROVED** [review]

File: v2/tools/generator/internal/config/property_configuration.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/config/property_configuration.go:30]: Done.

File: v2/tools/generator/internal/config/property_access.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		name,
+		property,
+		func(configuration *PropertyConfiguration) error {
+			c = a.accessor(configuration)
+			return nil
+		})
+
+	err := visitor.visit(a.model)
+	if err != nil {
+		// Something went wrong; we discard the error knowing that a
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/config/property_access.go:60]: Agreed. Done.

**Pull request merged**